core/src/main/scala/org/apache/spark/sql/execution/ddl.scala

```diff
@@ -367,16 +370,31 @@ case class SnappyCacheTableCommand(tableIdent: TableIdentifier,
       case None => session.catalog.cacheTable(tableIdent.quotedString)
       case Some(lp) =>
         val df = Dataset.ofRows(session, lp)
+        val isOffHeap = SnappyContext.getClusterMode(sparkSession.sparkContext)
+        match {
+          case _: ThinClientConnectorMode =>
+            SparkEnv.get.memoryManager.tungstenMemoryMode == MemoryMode.OFF_HEAP
+          case _ =>
+            try {
+              SnappyTableStatsProviderService.getService.getMembersStatsOnDemand.
+                  values.exists(member => (member.isDataServer
+                  && (member.getOffHeapMemorySize > 0)))
```

Comment:
I think we should check that all data servers should have off-heap, else persist with off-heap will fail if even one does not: values.forall(member => !member.isDataServer || member.getOffHeapMemorySize > 0)

Reply:
Changed..